### PR TITLE
Added comments to connection's implementation

### PR DESCRIPTION
**Pull request by bentoi**
This PR adds comments to Connection::message implementation.

### CODE REVIEW

File: cpp/src/Ice/ConnectionI.cpp [hunk 1]
```diff
@@ -1511,33 +1526,44 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
...
                     return;
                 }
 
+                // Validate the connection if it's not validate yet.
                 if (_state <= StateNotValidated && !validate(current.operation))
                 {
                     return;
                 }
 
+                // The connection is validated and doesn't need additional data to be read or written. So unregister
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionI.cpp:1565]: It's not clear to me what this means: is this the expected condition once the connection is validated?
But then, how do we receive the first incoming request on this connection?

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/Ice/ConnectionI.cpp [hunk 2]
```diff
@@ -576,7 +576,7 @@ Ice::ConnectionI::waitUntilFinished()
     // guarantee that there are no outstanding calls when deactivate()
     // is called on the servant locators.
     //
-    _conditionVariable.wait(lock, [this] { return _state >= StateFinished && _dispatchCount == 0; });
+    _conditionVariable.wait(lock, [this] { return _state >= StateFinished && _upcallCount == 0; });
```
[the remark this answers is not in the document]
Comment (bentoi) [reply cpp/src/Ice/ConnectionI.cpp:579]: I think it's about providing the guarantee that no more upcalls are running once the communicator destroy method returns.

File: cpp/src/Ice/ConnectionI.cpp [hunk 3]
```diff
@@ -1419,7 +1420,12 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
                 }
             }
 
-            while (readyOp & SocketOperationRead)
+            // If reads are ready, read the data into the connection's read buffer (_readStream). The data is read
```
[the remark this answers is not in the document]
Comment (bentoi) [reply cpp/src/Ice/ConnectionI.cpp:1423]: Yes, the sock/transceiver might be ready for writing and reading at the same time.

File: cpp/src/Ice/ConnectionI.cpp [hunk 4]
```diff
@@ -1511,33 +1526,44 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
...
 
+            // readOp and writeOp are set to the operations that the transport read or write calls from above returned.
+            // They indicate which operations will need to be monitored by the thread pool's selector when this method
+            // returns.
             SocketOperation newOp = static_cast<SocketOperation>(readOp | writeOp);
-            readyOp = static_cast<SocketOperation>(readyOp & ~newOp);
+
+            // Operations that are ready. For example, if message was called with SocketOperationRead and the transport
+            // read returned SocketOperationNone, reads are considered done: there's no additional data to read.
+            SocketOperation readyOp = static_cast<SocketOperation>(current.operation & ~newOp);
```
[the remark this answers is not in the document]
Comment (bentoi) [reply cpp/src/Ice/ConnectionI.cpp:1536]: I've removed the assert.

File: cpp/src/Ice/ConnectionI.cpp (hunk 1 above)
Comment (bentoi) [reply cpp/src/Ice/ConnectionI.cpp:1565]: The connection is in the holding state after being validated so we unregister the connection from the threadpool's selector. The connection will be activated by the connection factory (see activate calls on connections in `ConnectionFactory.cpp`).

Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp:1565]: Would be nice to insert this clarification in the new comment.

So when you create an OA, it accepts connections _before_ you activate this OA?
Whereas with IceRPC, a server only accepts connections after you call Listen:
https://docs.icerpc.dev/api/csharp/api/IceRpc.Server.html#IceRpc_Server_Listen

How does this work for new connections on an activated OA? Somehow after this line the OA actives the connection and re-register it with the thread pool?

And how does it work for outgoing connections?

Comment (bentoi) [reply cpp/src/Ice/ConnectionI.cpp:1568]: > Would be nice to insert this clarification in the new comment.

Fixed.

> So when you create an OA, it accepts connections before you activate this OA?

When you create the OA, an incoming connection factory is created for each endpoint. An incoming connection factory creates the transport acceptor and the acceptor starts listening for connections. So the OA does indeed starts listening for connections on creation but it doesn't accept them just yet.  Connections are only accepted once the `ObjectAdapter::activate` method is called. The OA activate ends up registering each endpoint server socket with the threadpool's selector.

> How does this work for new connections on an activated OA? Somehow after this line the OA actives the connection and re-register it with the thread pool?

Connections are accepted by `IncomingConnectionFactory::message`: https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L1405 This method is the same as for a connection, it's called by the thread pool when the TCP server socket is ready for accepting a connection.

This method creates the connection with the accepted transport transceiver. The connection is "started" which means it's initialized and then validated. Once done, the `IncomingConnectionFactory::connectionStartCompleted` is called:  https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L1612

The connection is activated there only if the OA is not in the hold state.

> And how does it work for outgoing connections?

It's always activated after the connection start (initialize + validate) completes: https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L891

**Pull request merged**